drivers/flash/Kconfig.mcux

```diff
@@ -41,6 +47,15 @@ config FLASH_MCUX_FLEXSPI_MX25UM51345G
 	select MEMC
 	select MEMC_MCUX_FLEXSPI
 
+config FLASH_MCUX_FLEXSPI_HYPERFLASH
+	bool "MCUX FlexSPI HYPERFLASH driver"
+	select FLASH_HAS_PAGE_LAYOUT
+	select FLASH_HAS_DRIVER_ENABLED
+	select MEMC
+	select MEMC_MCUX_FLEXSPI
+
+endchoice
+
```

Comment:
Can this be change to a menu instead of a choice. Suggestion is below:
```
menu "Flexspi flash driver (Select any one)"

config FLASH_MCUX_FLEXSPI_NOR
	bool "MCUX FlexSPI NOR driver"
	select FLASH_HAS_PAGE_LAYOUT
	select FLASH_HAS_DRIVER_ENABLED
	select MEMC
	select MEMC_MCUX_FLEXSPI

config FLASH_MCUX_FLEXSPI_MX25UM51345G
	bool "MCUX FlexSPI MX25UM51345G driver"
	select FLASH_HAS_PAGE_LAYOUT
	select FLASH_HAS_DRIVER_ENABLED
	select MEMC
	select MEMC_MCUX_FLEXSPI

config FLASH_MCUX_FLEXSPI_HYPERFLASH
	bool "MCUX FlexSPI HYPERFLASH driver"
	select FLASH_HAS_PAGE_LAYOUT
	select FLASH_HAS_DRIVER_ENABLED
	select MEMC
	select MEMC_MCUX_FLEXSPI

endmenu
```

Reply:
Using this menu should avoid changes to the `boards` folder for the various MXRT platforms

Reply:
> Can this be change to a menu instead of a choice. Suggestion is below:

Please read some of the history in this PR, then you'll see that you would end up with this situation:
https://github.com/zephyrproject-rtos/zephyr/pull/38676#discussion_r713773496

> In Kconfig I can select all of those simultaneously:
> ...
> but the CMake code here will only call zephyr_code_relocate() for one of them.
> I assume only one of those drivers should be allowed to select.
>
> Please fixup the Kconfig selection with proper depends (or consider a choice).

https://github.com/zephyrproject-rtos/zephyr/pull/38676#discussion_r717482440
> meaning that a user using zephyr/main can do:
>
> -   Disable all drivers
> -   Select multiple drivers (definitely an error)
>
> Now, the choice ensures that only a single driver can be selected.

Now, the solution you present here will allow a user to select more than one driver, which is clearly an error, especially if you look at current code on master.

So @Kludentwo is actually cleaning up existing mess in the Kconfig code by introducing a choice.

Reply:
I added a note in the title that suggests user to select one board. Making this a choice would require this move to the `_defconfig` file which we would like to avoid

Reply:
would you rather:
```
config FLASH_MCUX_FLEXSPI_HYPERFLASH
	bool "MCUX FlexSPI HYPERFLASH driver"
        depends on !FLASH_MCUX_FLEXSPI_NOR && !FLASH_MCUX_FLEXSPI_MX25

config FLASH_MCUX_FLEXSPI_MX25
	bool "MCUX FlexSPI MX25 driver"
        depends on !FLASH_MCUX_FLEXSPI_NOR && !FLASH_MCUX_FLEXSPI_HYPERFLASH

config FLASH_MCUX_FLEXSPI_NOR
	bool "MCUX FlexSPI NOR driver"
        depends on !FLASH_MCUX_FLEXSPI_MX25 && !FLASH_MCUX_FLEXSPI_HYPERFLASH
```
?

Reply:
> I added a note in the title that suggests user to select one board.

I noticed, and I disapprove.
Placing such info is not something we do elsewhere and this is why we have choice groups.
To ensure only a single choice item can be selected at a time.

> Making this a choice would require this move to the _defconfig file which we would like to avoid

Why ?
Please give a rationale why you would like to avoid this.
The `_defconfig` is a perfectly valid place for enabling of drivers for a board.

Reply:
> would you rather:

No.
That will give you a unresolvable dependency loop and make Kconfig fail.

If you try to fix that loop, then you will instead end up with a config like this, where not all options are visible to the user, like here:
![image](https://user-images.githubusercontent.com/9802583/135426069-430f55b3-494e-4069-bdf0-728646db10b2.png)

where the `MX25` and `HYPERFLASH` options are not visible until the NOR is deselected, as can be seen:
![image](https://user-images.githubusercontent.com/9802583/135426333-322ef051-d9f3-4e2c-aa79-9497bea48a80.png)

Reply:
A little late to the party, but isn't it perfectly valid to have both a FlexSPI NOR device and a FlexSPI hyperflash device connected for example?
The iMX RT1064 has 2 FlexSPI interfaces, where 1 is internally connected to the NOR device, On a custom board we connect the other FlexSPI to a hyperram device, but that could be a hyperflash.

However, where the execution code for the drive is located needs to be a single instance for the code relocation in case of XiP.

Reply:
> A little late to the party, but isn't it perfectly valid to have both a FlexSPI NOR device and a FlexSPI hyperflash device connected for example?

Theoretically it might be possible, but from my observations that I made in this comment I doubt it'll work correctly:
https://github.com/zephyrproject-rtos/zephyr/pull/38676#discussion_r713773496

That observation is based on this code:
https://github.com/zephyrproject-rtos/zephyr/blob/9448b68a6ce0f46a489f5cef6450e9240d8bd9c2/drivers/flash/CMakeLists.txt#L28-L36

as that relocation is guarded by `if()...elseif()...elseif()`.

So my main comment is that if they are mutual exclusive, a choice is the correct way to go.
If multiple can be selected, then the CMake code seems wrong.

I know you specifically referred to hyperflash together with nor device.
But the `MX25UM51345G` and `NOR` drivers are mutual exclusive as they defines the same functions and thus causes linker failures:
```
flash_mcux_flexspi_nor.c:26: multiple definition of `log_const_flash_flexspi_nor' .... flash_mcux_flexspi_mx25um51345g.c:26: first defined here
```

Reply:
I agree that `choice` is the right approach. I don't think we need the `optional` keyword. For cases where the user has a different Flash, they can add to this choice list and upstream the driver.

Reply:
> I don't think we need the optional keyword.

Great, if the `optional` can be dropped that will be fine with  me.
Just didn't wanted to ask a PR to remove such possibility without knowing if this was an intended possibility.

In which case there is no need to move to `_defconfig`.

Reply:
> I agree that `choice` is the right approach. I don't think we need the `optional` keyword. For cases where the user has a different Flash, they can add to this choice list and upstream the driver.

@mmahadevan108 the iMXRT1064-EVK even has the option to use both a hyperflash and the on-chip NOR flash, if you let this be a choice, you can't use both. For example, have the application running XiP on the NOR flash and having an NVS settings partition on the hyperflash.